corehq/apps/translations/views.py

```diff
@@ -0,0 +1,195 @@
+    def _uploaded_file_name(self):
+        uploaded_file = self.convert_translation_form.cleaned_data.get('upload_file')
+        return uploaded_file.name
+
+    @staticmethod
+    def _parse_excel_sheet(worksheet):
+        """
+        :return: the rows and the index for expected columns
+        """
+        rows = [row for row in worksheet.iter_rows()]
+        headers = [cell.value for cell in rows[0]]
+        source = headers.index('source')
+        translation = headers.index('translation')
+        occurrence = headers.index('occurrence') if 'occurrence' in headers else None
+        context = headers.index('context') if 'context' in headers else None
+        return rows, source, translation, occurrence, context
+
+    def _generate_translations_for_po(self, worksheet):
+        """
+        :return: a list of Translation objects
+        """
+        rows, source, translation, occurrence, context = self._parse_excel_sheet(worksheet)
+        translations = {worksheet.title: []}
+        for index, row in enumerate(rows[1:]):
+            _occurrence = row[occurrence].value if occurrence is not None else ''
+            _context = row[context].value if context is not None else ''
+            translations[worksheet.title].append(
+                Translation(
+                    row[source].value,
+                    row[translation].value,
+                    [(_occurrence, None)],
+                    _context)
+            )
+        return translations
+
+    def _generate_po_file(self, worksheet):
+        """
+        extract translations from worksheet and converts to a po file
+        :return: PoFileGenerator object
+        """
+        translations = self._generate_translations_for_po(worksheet)
+        po_file_generator = PoFileGenerator()
+        po_file_generator.generate_translation_files(translations, {})
+        return po_file_generator
+
+    def _generate_excel_file(self):
+        """
+        extract translations from po file and converts to a xlsx file
+        :return: Workbook object
+        """
+        uploaded_file = self.convert_translation_form.cleaned_data.get('upload_file')
+        po_file = polib.pofile(uploaded_file.read())
+        wb = openpyxl.Workbook()
+        ws = wb.worksheets[0]
+        ws.title = "Translations"
+        ws.append(['source', 'translation', 'context', 'occurrence'])
+        for po_entry in po_file:
+            ws.append([po_entry.msgid, po_entry.msgstr, po_entry.msgctxt,
+                       po_entry.occurrences[0][0] if po_entry.occurrences else ''])
+        return wb
+
+    def _po_file_response(self):
+        uploaded_file = self.convert_translation_form.cleaned_data.get('upload_file')
+        worksheet = openpyxl.load_workbook(uploaded_file).worksheets[0]
+        po_file_generator = self._generate_po_file(worksheet)
+        content = open(po_file_generator.generated_files[0][1], 'r', encoding="utf-8").read()
+        po_file_generator.cleanup()
```

Comment:
Seems like the cleanup should happen in a finally block. Also, proper file handling is a good idea. Two options:
```py
po_file_generator = self._generate_po_file(worksheet)
try:
    with open(po_file_generator.generated_files[0][1], 'r', encoding="utf-8") as fh:
        content = fh.read()
finally:
    po_file_generator.cleanup()
```
Or if you want to get more fancy you could make `po_file_generator` into a context manager and use it in a `with` block where `cleanup` gets called on `__exit__`.

Reply:
Okay. I never thought read operation would fail unless in some very unexpected scenario which would not be very harmful. Will add the block just to be on a safer side.